aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala

```diff
@@ -0,0 +1,114 @@
+import org.apache.commons.codec.binary.Base64
+import spray.json._
+
+import scala.util.Try
+
+
+sealed trait ModelSource {
+  def localVfs: File
+  def shouldDelete: Boolean
+}
+
+object ModelSource {
+
+  /**
+   * Allows for the following forms.
+   * {{{
+   * { "model":    "[Base 64 encoded byte array here]" }
+   * { "modelUrl": "some://apache.vfs2/url/here" }
+   * { "modelUrl": "some://apache.vfs1/url/here", "via": "vfs1" }
+   * { "modelUrl": "some://apache.vfs2/url/here", "via": "vfs2" }
+   * { "modelUrl": "/absolute/path/here",         "via": "file" }
+   * }}}
+   */
+  implicit object jsonFormat extends RootJsonFormat[ModelSource] with DefaultJsonProtocol {
+    override def read(json: JsValue) = {
+      val jso = json.asJsObject("Model Source expected to be object")
+
+      val modelVal = jso.getFields("model") match {
+        case Seq(JsString(m)) => Some(m)
+        case _                => None
+      }
+
+      val modelUrlVal = jso.getFields("modelUrl") match {
+        case Seq(JsString(m)) => Some(m)
+        case _                => None
+      }
+
+      // Default to VFS 2.
+      val vfsType = jso.getFields("via") match {
+        case Seq(via) => jso.convertTo(VfsType.JsonReader("via"))
+        case _        => VfsType.vfs2
+      }
+
+      val params = jso.getFields("params") match {
+        case Seq(p) => Option(p.convertTo[Either[Seq[String], String]])
+        case Nil    => None
+      }
+
+      val paramStr = params.map(_.fold(_.mkString(" "), identity)) getOrElse ""
+
+      val modelSource = (modelVal, modelUrlVal, vfsType) match {
+        case (None, Some(u), t)    => ExternalSource(Vfs.fromVfsType(t)(u))
+        case (Some(m), None, t)    => Base64StringSource(m, t)
+        case (Some(m), Some(u), _) => throw new DeserializationException("Exactly one of 'model' and 'modelUrl' should be supplied. Both supplied: " + json.compactPrint)
+        case (None, None, _)       => throw new DeserializationException("Exactly one of 'model' and 'modelUrl' should be supplied. Neither supplied: " + json.compactPrint)
+      }
+
+      modelSource
+    }
+
+    def modelFields(modelSource: ModelSource) = {
+      val ms = modelSource match {
+        case Base64StringSource(b64, t) => Seq("model" -> JsString(b64))
+        case ExternalSource(fs)         => Seq("modelUrl" -> JsString(fs.descriptor), "via" -> JsString(fs.vfsType.toString))
+      }
+      scala.collection.immutable.ListMap(ms:_*)
+    }
+
+    override def write(modelSource: ModelSource) = JsObject(modelFields(modelSource))
+  }
+}
+
+final case class Base64StringSource(b64EncodedData: String, vfsType: VfsType = VfsType.vfs2) extends ModelSource {
+
+  /**
+   * Get a Vfs instance pointing to a local file containing the base64-decoded b64EncodedData.
+   * @return
+   */
+  def localVfs = {
+    val localVfs = for {
+      tmpName <- getTmpFileName
+      v = getTmpVfs(tmpName)
+      _ <- copyToTmpFile(v)
```

Comment:
What does this do?  It doesn't look like it's actually writing anything to disk.